sync/src/types.rs

```diff
@@ -1067,6 +1085,82 @@ impl<CS: ChainStore> SyncSharedState<CS> {
         self.inflight_blocks.write().remove_by_peer(&pi);
         self.peers().disconnected(pi)
     }
+
+    pub fn insert_new_block(
+        &self,
+        chain: &ChainController,
+        pi: PeerIndex,
+        block: Arc<Block>,
+    ) -> bool {
+        let known_parent = |block: &Block| {
+            self.store()
+                .get_block_number(block.header().parent_hash())
+                .is_some()
+        };
+
+        // Insert the given block into orphan_block_pool if its parent is not found
+        if !known_parent(&block) {
+            debug!(
+                "insert new orphan block {} {:x}",
+                block.header().number(),
+                block.header().hash()
+            );
+            self.insert_orphan_block((*block).clone());
+            return false;
+        }
+
+        // Attempt to accept the given block if its parent already exist in database
+        if let Err(err) = self.accept_block(chain, pi, block.clone()) {
+            error!("accept block {:?} error {:?}", block, err);
+            return false;
+        }
+
+        // The above block has been accepted. Attempt to accept its descendant blocks in orphan pool.
+        // The returned blocks of `remove_blocks_by_parent` are in topology order by parents
+        let descendants = self.remove_orphan_by_parent(block.header().hash());
+        for block in descendants {
+            let block = Arc::new(block);
+
+            // If we can not find the block's parent in database, that means it was failed to accept
+            // its parent, so we treat it as a invalid block as well.
+            if !known_parent(&block) {
+                debug!(
+                    "parent-unknown orphan block, block: {}, {:x}, parent: {:x}",
+                    block.header().number(),
+                    block.header().hash(),
+                    block.header().parent_hash(),
+                );
+                continue;
+            }
+
+            if let Err(err) = self.accept_block(chain, pi, block.clone()) {
```

Comment:
avoids clone here, and remove the beginning line of loop: `let block = Arc::new(block);`
, change `debug` to`warn` and only output block hash.

Reply:
1. I will change the log as your comment.
2. Rust owner mechanism requires clone here if we wanna log `block_hash` within the error message. And this is Arc clone, it is cheap.